### PR TITLE
Handle missing devices in `/keys/claim` responses

### PR DESCRIPTION
**Pull request by richvdh**
Keep a record of devices that were included in a `/keys/claim` request, and then, if they are missing in the response, register them as "failed".

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/281.

~~This PR is currently based on https://github.com/matrix-org/matrix-rust-sdk/pull/2803.~~

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs [hunk 1]
```diff
@@ -44,6 +44,17 @@ use crate::{
...
+
+    /// If there is an active /keys/claim request, its details.
+    ///
+    /// This is used when processing the response, so that we can spot missing
+    /// users/devices.
+    ///
+    /// According to the doc on [`crate::OlmMachine::get_missing_sessions`],
+    /// there should only be one such request active at a time, so we only need
+    /// to keep a record of the most recent.
+    current_key_claim_request: Arc<StdRwLock<Option<(OwnedTransactionId, KeysClaimRequest)>>>,
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/session_manager/sessions.rs:56]: I think that this creates a harder to use API without giving any type of guidance to the user. It is adding some hidden state to the `OlmMachine` which the user needs to be aware of.

What does this offer compared to the suggestion posted here https://github.com/matrix-org/matrix-rust-sdk/issues/281#issuecomment-1564213984?

If we did what https://github.com/matrix-org/matrix-rust-sdk/issues/281#issuecomment-1564213984 suggests, the hidden state would not be hidden anymore, it would be part of the new struct, called `MissingSessions`. As I can tell the suggested approach has the following advantages:

1. No hidden state added to the `OlmMachine`.
2. The list of users would not need to be cloned and remembered in the `OlmMachine`.
3. A lock could be added to the new struct which would ensure that only one call to `get_missing_sessions()` can be active at the same time.
4. The type provides guidance on how to use the API, it clears up the constant question people have, if this response needs to be fed back into the machine or not. The only way to get rid of the type is to, either drop it, or call `MissingSessions::mark_request_as_sent()`.

I'm aware that we added the same type of hidden state to the `/keys/query` flows but there are some crucial differences:

1. The `/keys/query` flow happens automatically as part of the outgoing requests.
2. In the `/keys/query` flow we're not outright rejecting response because there's a mismatch, we're just not marking users as up-to-date.

That being said, I think that the outgoing requests flow could likely be improved with a similar new struct that remembers this state while the request is in flight instead of keeping hidden state in the `OlmMachine`.

Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:56]: Well, the honest answer here is that I hadn't seen https://github.com/matrix-org/matrix-rust-sdk/issues/281#issuecomment-1564213984 when I came up with this: tbh I find that issue a bit hard to follow as it seems to cover a range of different but related issues.

Still, I've had a look now. It seems like an interesting idea, but to address some of your points:

>  It is adding some hidden state to the `OlmMachine` which the user needs to be aware of.

Why does the user need to be aware of it? The user must already be aware that they can only make one call to `get_missing_sessions` at a time - the doc says that. Frankly I see this change as an improvement on that front because it will now warn you about it when you mess up.

> What does this offer compared to the suggestion posted here https://github.com/matrix-org/matrix-rust-sdk/issues/281#issuecomment-1564213984?

Two advantages to my approach, as I see it:

 * It keeps response handling consistent. Currently *all* HTTP responses, however the corresponding requests are generated, are fed back into `OlmMachine::mark_request_as_sent`. Making exceptions for individual requests does not make for an ergonomic API.

   I see your point that maybe the whole outgoing requests flow could be improved, and indeed if we considered similar changes to *all* outgoing requests, then that might be a good thing. To be honest, the whole `request_id` thing is a pain in the butt which would be nice to get rid of. But I don't think we should embark on such a project unless we have the time and determination to see it through properly, and I for one am not keen to take it on right now.

 * It means that no changes are needed on the application side. While it's true that we shouldn't tie ourselves in knots in an attempt to maintain a stable API, there is no doubt that changes to the API imply significant costs in terms of updating and testing downstream consumers. Frankly, one of the promises of the Rust Crypto SDK is that we can fix things once rather than 3 times. We ought to at least *consider* the costs of changing the API against the advantages of doing so.

And to answer your points about advantages of a separate `MissingSessions`.

> * No hidden state added to the `OlmMachine`.

I guess I don't really see why this is so bad. The term "hidden" seems quite loaded - as I see it, it's just *internal* state, and there's plenty of that already. Why is it better for `MissingSessions` to have internal state?

> * The list of users would not need to be cloned and remembered in the `OlmMachine`.

truuuue, but it will be remembered *somewhere* (in `MissingSessions`) and the only difference is that the clone can come a little later (in `MissingSessions::request()`). I'm not really convinced it's worthwhile.

> * A lock could be added to the new struct which would ensure that only one call to `get_missing_sessions()` can be active at the same time.

I don't really follow this. `get_missing_sessions` would be in `OlmMachine` and return a *new* `MissingSessions`, no? So the lock would need to be in `OlmMachine`. Which we could implement I guess...

> * The type provides guidance on how to use the API, it clears up the constant question people have, if this response needs to be fed back into the machine or not. The only way to get rid of the type is to, either drop it, or call `MissingSessions::mark_request_as_sent()`.

Yes, fair. Though again it speaks to: we could do with changing the API for *all* the outgoing requests. 

I guess what all of this comes down to is: if we want to change the API for outgoing requests, that might make sense, but I see it as something of a separate change. I don't think this PR makes it any harder to make that change, and now that I've got this far I'd kinda like to land what I've got rather than blocking it behind a refactorathon.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:56]: > Why does the user need to be aware of it? The user must already be aware that they can only make one call to `get_missing_sessions` at a time - the doc says that. Frankly I see this change as an improvement on that front because it will now warn you about it when you mess up.

That's partially true, you *should* make only one call to `get_missing_sessions()` with a overlapping set of users. Calling it with two disjoint sets of users is completely fine. Now even the main crate of the Rust SDK uses a crude lock and doesn't take advantage of this.

On the other hand if you do call it with two overlapping sets of users you will end up with multiple `Olm` sessions for the users/devices which are present in both sets.

After your patch, the two above scenarios will produce UTDs because we're refusing to create the `Olm` sessions and only a warning will be emitted in the logs, if people have logs enabled. I think this should at least be upgraded to an error, though it's a breaking change in the API as well.


> * It keeps response handling consistent. Currently _all_ HTTP responses, however the corresponding requests are generated, are fed back into `OlmMachine::mark_request_as_sent`. Making exceptions for individual requests does not make for an ergonomic API.

Fair point, but in any case we will change the API piece by piece so this temporary inconsistency is unavoidable.

>     * It means that no changes are needed on the application side. While it's true that we shouldn't tie ourselves in knots in an attempt to maintain a stable API, there is no doubt that changes to the API imply significant costs in terms of updating and testing downstream consumers. Frankly, one of the promises of the Rust Crypto SDK is that we can fix things once rather than 3 times. We ought to at least _consider_ the costs of changing the API against the advantages of doing so.

I don't think that's necessarily true, it might be true for all the places where we are using the crypto crate, but as stated above your change is breaking things as well.

> And to answer your points about advantages of a separate `MissingSessions`.
> 
> > * No hidden state added to the `OlmMachine`.
> 
> I guess I don't really see why this is so bad. The term "hidden" seems quite loaded - as I see it, it's just _internal_ state, and there's plenty of that already. Why is it better for `MissingSessions` to have internal state?

I think it's better to put it into `MissingSessions` because the lifetime of that state is tied to this struct. We don't permanently change the state of the `OlmMachine` until you remember to mark the current request as sent or call `get_missing_sessions()` again. 

And the hidden state will prevent you from creating `Olm` sessions unless those two calls match up again. Granted the scenario sounds like a bug that should be fixed in the application, but this API seems to be much more fragile because of the newly introduced state. Again, the misuse of the API now produces UTDs, not duplicate Olm sessions like before.

> truuuue, but it will be remembered _somewhere_ (in `MissingSessions`) and the only difference is that the clone can come a little later (in `MissingSessions::request()`). I'm not really convinced it's worthwhile.

After rethinking this, it would probably be too hard to avoid cloning this, certainly in any FFI crate where borrowing usually doesn't work.

> I don't really follow this. `get_missing_sessions` would be in `OlmMachine` and return a _new_ `MissingSessions`, no? So the lock would need to be in `OlmMachine`. Which we could implement I guess...

Yes, that's correct. `MissingSession` itself would not be in `OlmMachine`, it would be created when you call `get_missing_sessions()`. It would contain a `MutexGuard`, while the `OlmMachine` would contain the `Mutex<()>`.

> Yes, fair. Though again it speaks to: we could do with changing the API for _all_ the outgoing requests.

Yes, but as I said, this would need to be gradual.
 
> I guess what all of this comes down to is: if we want to change the API for outgoing requests, that might make sense, but I see it as something of a separate change. I don't think this PR makes it any harder to make that change, and now that I've got this far I'd kinda like to land what I've got rather than blocking it behind a refactorathon.

Well perhaps we don't want to tackle changing the API to the suggested pattern right now, that's true. It still seems that the change in this PR is problematic.

Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:56]: Ok, as we discussed, I have reworked this so that it will fall back gracefully if it cannot find details of the request: it will skip the whole `failed_devices` bit in that case. Please take another look?

**Review by poljar — APPROVED** [review]
Looks good, left a couple of suggestions.

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs [hunk 2]
```diff
@@ -336,71 +379,84 @@ impl SessionManager {
...
-        // process all the missing devices at once to save repeatedly grabbing the lock
-        if !missing_devices.is_empty() {
-            warn!(
-                ?missing_devices,
-                "Tried to create a new sessions, but the signed one-time key was missing for some devices",
-            );
+        // if we were able to pair this response with a request, look for devices that
+        // were present in the request but did not elicit a successful response.
+        if let Some(request) = request {
+            let missing_devices: Vec<(OwnedUserId, OwnedDeviceId)> = request
```
[the remark this answers is not in the document]
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:414]: ok, happy to do this. The reason I avoided it was that it felt a bit redundant to build the two sets just so that we could call `.difference` on them, particularly given we already have the `requested_sessions` in a map whose keys we can check in O(n log n) (ie, the same as the `BTreeSet`).

I guess using sets shows the intention more clearly though.

[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs]: Luckily we can check if the `BTreeSet` approach is slower, red is the `flat_map()`, blue is the `BTreeSet` approach:

flat_map_vs_sets_page-0001

The benchmark is processing a `/keys/claim` response with 163 one-time keys. I did run it a couple of times to get somewhat stable results. `BTreeSet` seems to be either a bit better or at worst it makes no difference which approach we pick.

[the remark this answers is not in the document]
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs]: To check I am understanding the report correctly: this is saying the mean time increased by 0.7680%? In which case, very much agreed this is not worth worrying about.

I'd be interested to see how you ran this benchmark, for future reference.

[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs]: You are understanding correctly, I did have a couple of runs which claimed that the time decreased by ~0.5% as well. As the report tells us, all of this is within the noise threshold.

> I'd be interested to see how you ran this benchmark, for future reference.

Please take a look at the `/benchmarks` folder:  https://github.com/matrix-org/matrix-rust-sdk/tree/main/benchmarks
